### DISCUSSION TITLE
[A-L] (23 ottobre 2023) Esercizio 4

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#82] [Esami A-L] [answered]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/938e969c-2625-4ead-9323-71a854b2d95b)

**Comment by Frisbon** [comment, marked as answer]
```python
A = [1,2,3,5,7,8,9] #mi da 4
A = [1,2,4,5,7,8,9,10,11,12,14] #mi da 3
A = [4,5,7,8,9,10,11]  #mi da 6

i = 0; j = len(A)-1
minimo = 99999
while (i<j):
    m = (i+j)//2
    
    #se a destra ho minchione
    if (m+1<len(A)) and (A[m] != A[m+1]-1):
        possibile = A[m]+1 
        minimo = min(minimo , possibile)
            
        #se a sinistra ho minchione
    if (0<=m-1) and (A[m] != A[m-1]+1):
        possibile = A[m]-1 
        minimo = min(minimo , possibile)
    
    if m != A[m] - 1:
        j = m-1
    else:
        i = m+1

print(minimo)
```

**Reply by Frisbon** [reply]
La fun part begins se i get steccato
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/84680a07-e490-4215-99ea-032e749ae6e2)
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/1591e971-a0fc-4fa3-9832-a20caf870d4a)

**Reply by Elia-Belli** [reply]
Nel terzo esempio:
`A = [4,5,7,8,9,10,11]`
la soluzione non dovrebbere essere $x = 1$ ? 
Siccome si tratta del minimo numero intero positivo $x\gt 0$ non contenuto in $A$

**Reply by Blackcat765** [reply]
per A=[3,4,5,6,7,8,10] da come risultato 99999,quando il risultato dovrebbe essere 9 , presumo sia colpa delle ultime 4 righe, in particola modo " if m != A[m] - 1: "

**Reply by Blackcat765** [reply]
se le cambi con


       offset=A[0]-1
        if A[m] == m + 1 + offset:
            i = m+1
        else:
            j = m-1

usate da Marco funziona

**Comment by MarcoLinardi** [comment]
```
def num_mancante(A)
	i = 0
	j = len(A)-1
	offset = A[0] - 1

	while i < j
                m = (i + j) / 2
		if A[m] == m + 1 + offset
			i = m + 1
		else
			j = m 

	return A[i] - 1
```
Ragiono sul fatto che essendo ordinato e cercando il minimo numero mancante, avremo un valore v nell'indice v dell'array, utilizzando un offset per controllare gli indici nel caso in cui il primo valore sia diverso da 1.

**Reply by Frisbon** [reply]
Ho testato il tuo codice e sembra ci sia qualche bug

```py
def num_mancante(A):
    i = 0
    j = len(A)-1
    offset = A[0] - 1
    while i<j:
        m = (i+j)//2
        if A[m] == m+1+offset: i = m+1
        else: j = m-1
    return i-1 +offset

A = [1,2,3,5,7,8,9] #manca 4
print(num_mancante(A))
A = [1,2,4,5,7,8,9,10,11,12,14] #manca 3
print(num_mancante(A))
A = [4,5,7,8,9,10,11]  #manca 6           
print(num_mancante(A))

---------
RESTITUISCE
1
0
4

**Reply by MarcoLinardi** [reply]
ho modificato alcuni errori che avevo scritto nel codice, non avevo testato

**Reply by LAMASE** [reply]
L'esercizio chiede il min x > 0 che non appartiene ad A, quindi l'offset non è richiesto: se hai "4,5,6" la risposta dovrebbe essere 1.
Non è neanche detto che il "buco" sia di un solo numero: se hai "1,2,5", dopo aver testato che il 2 è nella posizione giusta e vai a destra sul solo 5, non devi restituire A[i] - 1 (che sarebbe 4) ma 3. Meglio A[i-1] + 1.

Aggiungerei quindi un controllo sul primo elemento, toglierei l'offset e restituirei A[i-1] + 1 (esiste sempre l'indice i-1 mettendo il controllo iniziale)

**Comment by Warcophyr** [comment]
```Python
def es4_ric(A, i, j):
    if (j-i-1) == 1 or (j-i-1) == 0:
        if A[j] != j+1:
            return j+1
        else:
            return float("inf")
    m = (i+j) // 2

    if A[m] != (m-i+1):
        return min(A[m]-1, es4_ric(A, i, m-1))
    else:
        return min(float("inf"), es4_ric(A, m+1, j))
    

def es4(A):
    i = 0
    j = len(A)
    res = es4_ric(A, i, j)
    if res < A[0]:
        return 1
    else:
        return res


if __name__ == "__main__":
    A = [1, 2, 3, 5, 7, 9]
    B = [1, 2, 4, 5, 7, 8, 9, 10, 11, 12, 14]
    C = [4, 5, 7, 8, 9, 10, 11]

    print(es4(A))
    print(es4(B))
    print(es4(C))
```

**Comment by Elia-Belli** [comment]
## Divide et Impera
Modifichiamo la ricerca binaria per rientrare nella complessità $O(\log n)$, abbiamo alcuni casi particolari:
- Se $A[0] \gt 1$ la soluzione è $1$, in quanto minimo numero positivo $\not \in A$
- Altrimenti, se in $A$ tutti gli elementi sono consecutivi, ovvero $A[i+1] = A[i] + 1$ , la soluzione è $A[n] + 1$
- Altrimenti dobbiamo ricercare un buco nell'array, favorendo la parte sinistra in quanto cerchiamo il minimo.

```python
def buco(A):
    start = 0
    end = len(A) - 1
	
    if(A[0] > 1): return 1
    if(A[n] - A[0] > len(A)): return A[n] + 1

    while(end - start > 1):
	m = (end+start)/2
	if(A[m] - A[start] > m - start):
		end = m
	elif(A[end] - A[m] > end - m):
		start = m

    return A[start] + 1
```

**Comment by IonutCicio** [comment]
```rust
fn smallest_search(array: &[usize]) -> usize {
    if array[0] != 1 {
        return 1;
    }

    let mut left = 0;
    let mut right = array.len() - 1;

    while left < right - 1 {
        let middle = (left + right) / 2;

        if array[middle] == middle + 1 {
            left = middle;
        } else {
            right = middle;
        }
    }

    array[left] + 1
}
```

**Comment by CiottoloMaggico** [comment]
![immagine](https://github.com/user-attachments/assets/b30e3714-3d1b-4f57-95a8-e7fdf9c1d7ae)

**Comment by FeddyLix17** [comment]
Per avere complessità $\Large O(log\ n)$ è possibile fare uso della [ricerca binaria](https://en.wikipedia.org/wiki/Binary_search#:~:text=Binary%20search%20begins%20by%20comparing,lower%20half%20of%20the%20array.).

```python3
def es4(A):
    # se il primo elemento è maggiore di 1, il minimo intero escluso è 1
    if A[0] > 1:
        return 1

    left, right = 0, len(A) - 1

    while left <= right:
        mid = (left + right) // 2
        
        # se A[mid] è uguale a mid + A[0], x non è presente fino a quel punto
        if A[mid] == mid + A[0]:
            left = mid + 1
        # altrimenti, x non si troverà in nessun elemento successivo a A[mid]
        else:
            right = mid - 1

    # x sarà left + 1, considerato che il primo elemento ha indice 0
    return left + 1
```

**Comment by niccolo-fato** [comment]
```py
def esercizio_4(A):
    left = 0
    right = len(A) - 1

    while left <= right:
        mid = (left + right) // 2

        if A[mid] != mid + 1:
            right = mid - 1 
        else:
            left = mid + 1   
    return left + 1

```

**Comment by cek753** [comment]
<img width="1320" height="785" alt="image" src="https://github.com/user-attachments/assets/96016ca0-bcf3-45db-ad8c-85b1eddeeab9" />
